Update to use C++20 std library features

abigailfox opened the pull request:
#### b4c015abf2c705e0c79a32d0090da9be6a723959
<pre>
Update to use C++20 std library features
<a href="https://bugs.webkit.org/show_bug.cgi?id=280808">https://bugs.webkit.org/show_bug.cgi?id=280808</a>

Reviewed by Abrar Rahman Protyasha, Sam Weinig, and Mike Wyrzykowski.

Now that C++20 is required, we can take advantage of new standard library features and simplify some code. This addresses several FIXMEs from before we required C++20

* Source/WTF/wtf/SortedArrayMap.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isSortedConstExpr): Deleted.
(WTF::allOfConstExpr): Deleted.
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::tableForEncoding):
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::mergeFeatures):
(WebGPU::includesUnsupportedFeatures):

Canonical link: <a href="https://commits.webkit.org/284735@main">https://commits.webkit.org/284735@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5648b83f16a7d5cf2f414fffc4000b946fe204d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69978 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49379 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74063 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21136 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72095 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Checked out pull request; Passed bindings tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57179 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20987 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55525 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14009 "Passed tests") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60352 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36006 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41643 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; Running layout-tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19513 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63094 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63565 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18131 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Running checkout-pull-request; Skipped layout-tests; 9 flakes") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75778 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69224 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14203 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17358 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63223 "Passed tests") | 
| | [⏳ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60427 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63162 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15586 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11180 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4788 "Build is in progress. Recent messages:OS: Ventura (13.6.9), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 11 flakes") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91006 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45182 "Built successfully") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16171 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Found 100 jsc stress test failures: ChakraCore.yaml/ChakraCore/test/Function/prototype.js.default, ChakraCore.yaml/ChakraCore/test/Operators/instanceof.js.default, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout-dfg-eager-no-cjit, jsc-layout-tests.yaml/js/script-tests/exception-instanceof.js.layout-no-cjit ..., JSC test binary failures: testb3, testapi; Compiled JSC; Running jscore-test-without-change; Passed JSC tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46256 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47527 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45997 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/SortedArrayMap.h:
```diff
@@ -159,7 +159,7 @@ template<ASCIISubset subset, typename CharacterType> constexpr std::make_unsigne
     : literal { ASCIILiteral::fromLiteralUnsafe(characters) }
 {
-    ASSERT_UNDER_CONSTEXPR_CONTEXT(allOfConstExpr(&characters[0], &characters[size - 1], [] (char character) {
+    ASSERT_UNDER_CONSTEXPR_CONTEXT(std::all_of(&characters[0], &characters[size - 1], [] (char character) {
```
Might be able to use std::ranges::all_of(characters, [] …) to make this a bit shorter (can’t recall off hand if it works with that construction).

weinig replied:
(Generally, the ranges versions can probably be used in the PR).

aprotyas replied:
I was going to say `std::ranges::all_of` is missing from pre-Xcode 14.3, but turns out our oldest target is building against Xcode 14.3, so we're golden 😄

abigailfox replied:
I actually don't think I can do this in these asserts it turns out, since `std::range::all_of` returns a `constexpr bool` and the ranges version fails with `non-constexpr function 'WTFReportAssertionFailure' cannot be used in a constant expression`. The regular `std::all_of` returns just a regular `bool` and compiles fine in the assert (this is also true for `is_sorted`).

That said, if we have a preference for the `std::ranges` syntax since it looks cleaner, there's several other places where it could be swapped in and the `constexpr` difference shouldn't matter.

weinig reviewed (approved):
I realize this is still technically a draft, but looks good to me.

abigailfox committed:
Update to use C++20 std library features
https://bugs.webkit.org/show_bug.cgi?id=280808

Reviewed by Abrar Rahman Protyasha, Sam Weinig, and Mike Wyrzykowski.

Now that C++20 is required, we can take advantage of new standard library features and simplify some code. This addresses several FIXMEs from before we required C++20

* Source/WTF/wtf/SortedArrayMap.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::isSortedConstExpr): Deleted.
(WTF::allOfConstExpr): Deleted.
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::tableForEncoding):
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::mergeFeatures):
(WebGPU::includesUnsupportedFeatures):

Canonical link: https://commits.webkit.org/284735@main
